### PR TITLE
Refactor ingest main work loop

### LINKED ISSUE #1519: ingest loops when re-queueing messages
When trying to reingest a file through restoring a message, ingest does not check if the file is ingested so it is returned back to the queue and we end up with an stuck message in the archived.

We need some extra checks before the file event creation
Ingest should check the last status from the `file_event_log` and if the work that the service is supposed to do has happened, then the file should be moved along to the next service.

**How to reproduce:**
Send an ingest message to an already Ingested file.
We should not end up with messages stuck in a loop.

**Acceptance criteria:**

- [ ] If the message is re-queued, we should log the event/error
- [ ] Tests, if possible

**Additional context**
Created from #1463. Also see #1377, #1505 and #1082.
Also check if #1037 can be closed when this is fixed.

### LINKED ISSUE #1561: [ingest] break apart the main work loop
## Please describe the feature

As a developer, I want all the code to statically testable so that It is easier to create a full test suite even for extreme edge cases.

## Acceptance criteria

- [ ] The main function send each message to a worker function that returns ack or nack with regards to the message.
- [ ] Testsuite is created
- [ ] All tests pass

## Estimation of size

small/medium/big

## Estimation of priority

low/medium/high

### PR DESCRIPTION
**Pull request by jbygdell**
## Related issue(s) and PR(s)
This PR closes #1561 

This PR also closes #1519 and #1037.

## Description
The code for the main pipeline components are all cluttered i n the main function. That makes it all but impossible to run any tests other than full integration tests using the final container.

This PR breaks apart the main work loop in `ingest` into parts that can be tested and adds test for the following conditions:
* TestCancelFile
* TestCancelFile_wrongCorrelationID
* TestIngestFile
* TestIngestFile_missingFile
* TestIngestFile_reingestCancelledFile
* TestIngestFile_reingestCancelledFileNewChecksum
* TestIngestFile_reingestVerifiedCancelledFile
* TestIngestFile_reingestVerifiedCancelledFileNewChecksum
* TestIngestFile_reingestVerifiedFile
* TestIngestFile_secondTime
* TestIngestFile_unknownInboxType

## How to test
While in the `sda` folder run `go test -v cmd/ingest/*.go` to only test the ingest part. In order to test all of the GO code just run `go test`.

### CODE REVIEW

File: sda/cmd/verify/verify.go [hunk 1]
```diff
@@ -392,6 +392,15 @@ func main() {
 					continue
 				}
 
+				if err := db.UpdateFileEventLog(message.FileID, "verified", delivered.CorrelationId, "ingest", "{}", string(verifiedMessage)); err != nil {
```
Comment (MalinAhlberg) [inline sda/cmd/verify/verify.go:395]: why is this added..?

File: sda/internal/database/db_functions.go [hunk 2]
```diff
@@ -202,34 +202,37 @@ func (dbs *SDAdb) storeHeader(header []byte, id string) error {
...
+		if err == nil {
+			break
+		}
+		time.Sleep(time.Duration(math.Pow(2, float64(count))) * time.Second)
 	}
 
 	return err
 }
-func (dbs *SDAdb) setArchived(file FileInfo, fileID, corrID string) error {
+func (dbs *SDAdb) setArchived(file FileInfo, fileID string) error {
```
Comment (MalinAhlberg) [inline sda/internal/database/db_functions.go:218]: no longer updates the `file_event_log`?

File: sda/cmd/ingest/ingest.go [hunk 3]
```diff
@@ -137,440 +149,445 @@ func main() {
...
-
-					// Restart on new message
-					continue
-				}
+func (app *Ingest) cancelFile(correlationID string, message schema.IngestionTrigger) string {
+	fileUUID, err := app.DB.GetFileID(correlationID)
+	if err != nil {
+		log.Errorf("failed to get ID for file from message: %s, reason: %s", correlationID, err.Error())
+		if strings.Contains(err.Error(), "sql: no rows in result set") {
+			return "reject"
```
Comment (MalinAhlberg) [inline sda/cmd/ingest/ingest.go:205]: I don't follow the logic of `reject`s... Could you maybe add something in the PR description or in comments, or in a stand-up?

**Review by MalinAhlberg — COMMENTED** [review]
I think the ingest loop refactoring is very nice :star:!

The whole refactoring is a bit hard to follow though, so sorry for the confused questions. There seems to be a bit more than just refactoring going on, and I don't really follow. So I put my initial questions down in the comments.

File: sda/cmd/ingest/ingest.go [hunk 4]
```diff
@@ -137,440 +149,445 @@ func main() {
...
-					if err != nil {
-						log.Errorf("failed to get ID for file, reason: %s", err.Error())
-						if err := delivered.Nack(false, true); err != nil {
-							log.Errorf("failed to Nack message, reason: (%s)", err.Error())
-						}
-
-						continue
-					}
+			default:
+				log.Errorln("better message needed")
```
Comment (nanjiangshu) [inline sda/cmd/ingest/ingest.go:189]: Will it be better to print the content of the message? Otherwise it is not easy to debug why it is failed. Also, it is not very clear what does `better message` mean.

File: sda/cmd/ingest/ingest.go [hunk 5]
```diff
@@ -137,440 +149,445 @@ func main() {
...
-						log.Errorf("failed to publish message, reason: (%s)", err.Error())
-					}
-
-					// Restart on new message
-					continue
-				}
+func (app *Ingest) cancelFile(correlationID string, message schema.IngestionTrigger) string {
+	fileUUID, err := app.DB.GetFileID(correlationID)
+	if err != nil {
+		log.Errorf("failed to get ID for file from message: %s, reason: %s", correlationID, err.Error())
```
Comment (nanjiangshu) [inline sda/cmd/ingest/ingest.go:203]:
```suggestion
		log.Errorf("failed to get ID for file from message (correlationID: %s), reason: %v", correlationID, err)
```

Comment (nanjiangshu) [reply sda/cmd/ingest/ingest.go:203]: Otherwise, it is hard to know the value printed is correlationID without looking at the code.

File: sda/cmd/ingest/ingest.go (hunk 3 above)
Comment (nanjiangshu) [reply sda/cmd/ingest/ingest.go:205]: I'm also not super clear about the new logic, but it looks quite neat.

It seems each action will return one of the three status 
- `ack`
- `nack`
- `reject`

It it works well, maybe we can apply this logic to other services as well, or?

Comment (jbygdell) [reply sda/cmd/ingest/ingest.go:205]: There are 3 responses to a message:
Ack - All ok
Nack - Failure, re queue message and try again
Reject - failure, don't re queue the message. Could as well has been Ack but used to be able to tell good  from bad results in the tests among other things. For example an error where the message is not to be re queued.

File: sda/cmd/ingest/ingest.go (hunk 4 above)
Comment (jbygdell) [reply sda/cmd/ingest/ingest.go:189]: Did I forget to change that? It should hold a relevant message.

File: sda/internal/database/db_functions.go (hunk 2 above)
Comment (jbygdell) [reply sda/internal/database/db_functions.go:218]: If you look in ingest.go where this is called from you will see that the file_event_log update is done separately.

File: sda/cmd/ingest/ingest.go [hunk 6]
```diff
@@ -137,440 +148,403 @@ func main() {
...
-					body, _ := json.Marshal(fileError)
-					if err = mq.SendMessage(delivered.CorrelationId, conf.Broker.Exchange, "error", body); err != nil {
-						log.Errorf("failed to publish message, reason: (%s)", err.Error())
-					}
-
-					// Restart on new message
-					continue
-				}
+func (app *Ingest) cancelFile(correlationID string, message schema.IngestionTrigger) string {
+	fileUUID, err := app.DB.GetFileID(correlationID)
```
[the remark this answers is not in the document]
Comment (jbygdell) [reply sda/cmd/ingest/ingest.go:201]: That should not be possible without an DB error.

File: sda/internal/database/db_functions.go [hunk 7]
```diff
@@ -295,35 +298,42 @@ func (dbs *SDAdb) getHeader(fileID string) ([]byte, error) {
...
-		file.DecryptedChecksum,
-		"SHA256",
-	)
+	const verified = "UPDATE sda.files SET decrypted_file_size = $1 WHERE id = $2;"
+	if _, err := dbs.DB.Exec(verified, file.DecryptedSize, fileID); err != nil {
+		return fmt.Errorf("setVerified error: %s", err.Error())
+	}
 
-	return err
+	const addArchiveChecksum = "INSERT INTO sda.checksums(file_id, checksum, type, source)" +
```
Comment (nanjiangshu) [inline sda/internal/database/db_functions.go:322]: I find long SQL code like this is not easy to read when using string concatenation like this. I think using raw string is easier to read and maintain and also easier to copy to SQL shell for manual debugging, for example

```golang
const addArchiveChecksum = `
INSERT INTO sda.checksums(file_id, checksum, type, source)
VALUES ($1, $2, upper($3)::sda.checksum_algorithm, upper('ARCHIVED')::sda.checksum_source)
ON CONFLICT ON CONSTRAINT unique_checksum
DO UPDATE SET checksum = EXCLUDED.checksum;
`
```

What do you think? Should we fixed in this refactoring PR or should I create a new issue for that?

**Review by nanjiangshu — COMMENTED** [review]
I tested manually with repeated ingestion and the errors described in the issues are gone. Great work! Apart from that, I added a few minor comments.

File: sda/cmd/verify/verify.go (hunk 1 above)
Comment (jbygdell) [reply sda/cmd/verify/verify.go:395]: Because `setVerified` is moved to GO and to keep all calls identical in `s3inbox`, `ingest` and `verify`

This change is brought by the fact that ingest now can handle cases correctly, re-ingesting a file where a new version is uploaded to the inbox, with new checksum.

File: sda/internal/database/db_functions.go (hunk 7 above)
Comment (jbygdell) [reply sda/internal/database/db_functions.go:322]: In that case we probably should do a dedicated rework of all SQL statements as a separate PR since the current style is used elsewhere.

File: sda/cmd/ingest/ingest.go (hunk 4 above)
Comment (jbygdell) [inline sda/cmd/ingest/ingest.go:189]:
```suggestion
```

File: sda/cmd/ingest/ingest.go [hunk 8]
```diff
@@ -137,440 +149,445 @@ func main() {
...
+	}
+	readBuffer := make([]byte, bufSize)
+	hash := sha256.New()
+	var bytesRead int64
+	var byteBuf bytes.Buffer
+
+	for bytesRead < fileSize {
+		i, _ := io.ReadFull(file, readBuffer)
+		if i == 0 {
+			log.Debugln("this should not happen")
```
Comment (MalinAhlberg) [inline sda/cmd/ingest/ingest.go:404]: An error log would be useful here..?

File: sda/cmd/ingest/ingest.go (hunk 3 above)
Comment (MalinAhlberg) [reply sda/cmd/ingest/ingest.go:204]: Ok, I see, it sounds reasonable to me. I added a suggestion of a comment to the code above, for clarity for other confused developers.

File: sda/internal/database/db_functions.go (hunk 7 above)
Comment (nanjiangshu) [reply sda/internal/database/db_functions.go:322]: I created a new issue https://github.com/neicnordic/sensitive-data-archive/issues/1584 for it.

File: sda/cmd/ingest/ingest.go [hunk 9]
```diff
@@ -137,440 +148,403 @@ func main() {
...
+		}
+	case "":
+		// Catch all for inboxes that doesn't update the DB
+		fileID, err = app.DB.RegisterFile(message.FilePath, message.User)
+		if err != nil {
+			log.Errorf("InsertFile failed, reason: (%s)", err.Error())
+
+			return "nack"
+		}
+	case "uploaded":
```
[the remark this answers is not in the document]
Comment (MalinAhlberg) [reply sda/cmd/ingest/ingest.go:328]: I see, thanks.
But there is no case for `enabled` now?

File: sda/cmd/ingest/ingest.go [hunk 10]
```diff
@@ -137,440 +149,444 @@ func main() {
...
+	}
+	readBuffer := make([]byte, bufSize)
+	hash := sha256.New()
+	var bytesRead int64
+	var byteBuf bytes.Buffer
+
+	for bytesRead < fileSize {
+		i, _ := io.ReadFull(file, readBuffer)
+		if i == 0 {
+			log.Debugln("this should not happen")
```
Comment (jbygdell) [inline sda/cmd/ingest/ingest.go:404]:
```suggestion
			log.Errorln("readBuffer returned 0 bytes, this should not happen")
```

**Review by MalinAhlberg — APPROVED** [review]
I also tested to re-ingest, no loops occur anymore. Great work!

**Review by nanjiangshu — APPROVED** [review]
Looks good.
